File: modules/analysis/templates/02-clean-describe/cleaning.py
#!/usr/bin/env python3
"""Read, clean, describe, and save analysis data."""

from __future__ import annotations

import argparse
import sys
from datetime import date
from pathlib import Path


def _load_common():
    for parent in Path(__file__).resolve().parents:
        if (parent / "_shared").exists():
            sys.path.insert(0, str(parent / "_shared"))
            import python_common  # type: ignore

            return python_common
    raise RuntimeError("Cannot locate templates/_shared/python_common.py")


common = _load_common()


def main() -> int:
    parser = common.add_common_args(argparse.ArgumentParser(description=__doc__))
    args = parser.parse_args()
    p = common.paths(args.out_root)
    if not args.data or not Path(args.data).exists():
        report = common.write_markdown(p["reports"] / f"cleaning-report-{date.today().isoformat()}.md", "Cleaning Report", {"阻断": "缺少 --data 或数据文件不存在。"})
        common.log_run(args, "02-clean-describe", "blocked", [report], "缺少可执行数据输入。")
        return 2
    raw = common.load_data(args.data)
    cleaned = common.clean_data(raw)
    roles = common.infer_roles(cleaned, args.dict_path)
    dict_path = common.write_variable_dictionary(cleaned, roles, args.data, args.out_root)
    sample, sample_flow = common.sample_flow(cleaned, roles, args.out_root)
    data_path = common.save_analysis_data(sample, args.out_root)
    table1 = common.describe_data(sample, [roles.y, roles.x, *roles.controls], args.out_root)
    corr = common.correlation_table(sample, [roles.y, roles.x, *roles.controls], args.out_root)
    figs = common.basic_figures(sample, roles, args.out_root)
    report = common.write_markdown(
        p["reports"] / f"cleaning-report-{date.today().isoformat()}.md",
        "Cleaning Report",
        {
            "数据输入": str(args.data),
            "样本": f"raw N={len(raw)}; analysis N={len(sample)}",
            "变量角色": str(roles),
            "清洗规则": "列名标准化、特殊缺失码转 NA、数值变量 1%/99% 缩尾、去重、删除 Y/X/control 缺失。",
        },
    )
    outputs = [data_path, dict_path, sample_flow, table1, report, *(x for x in [corr] if x), *figs]
    common.log_run(args, "02-clean-describe", "ok", outputs, "已真实读取、清洗并生成描述统计。")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: modules/analysis/templates/02-clean-describe/cleaning.R
# Read, clean, describe, and save analysis data.

`%||%` <- function(a, b) if (is.null(a) || length(a) == 0) b else a
find_shared <- function() {
  cmd <- commandArgs(FALSE); file_arg <- sub("^--file=", "", cmd[grep("^--file=", cmd)][1] %||% "")
  p <- if (nzchar(file_arg)) dirname(normalizePath(file_arg, mustWork = FALSE)) else getwd()
  repeat { candidate <- file.path(p, "_shared", "r_common.R"); if (file.exists(candidate)) return(candidate); np <- dirname(p); if (identical(np, p)) stop("Cannot locate templates/_shared/r_common.R"); p <- np }
}
source(find_shared())
args <- parse_common_args(); ensure_dirs(args$out_root)
if (is.na(args$data) || !file.exists(args$data)) {
  report <- write_md(file.path(args$out_root, "reports", paste0("cleaning-report-", Sys.Date(), ".md")), "Cleaning Report", c("阻断" = "缺少 --data 或数据文件不存在。"))
  log_run(args, "02-clean-describe", "blocked", report, "缺少可执行数据输入。")
  quit(status = 2)
}
raw <- load_data(args$data)
cleaned <- clean_data(raw)
roles <- infer_roles(cleaned, args$dict)
dict_path <- write_variable_dictionary(cleaned, roles, args$data, args$out_root)
sample <- sample_flow(cleaned, roles, args$out_root)
data_path <- file.path(args$out_root, "data", "analysis-data.csv")
write_csv_safe(sample, data_path)
table1 <- make_table1(sample, c(roles$y, roles$x, roles$controls), args$out_root)
report <- write_md(file.path(args$out_root, "reports", paste0("cleaning-report-", Sys.Date(), ".md")),
                   "Cleaning Report",
                   c("数据输入" = args$data,
                     "样本" = paste0("raw N=", nrow(raw), "; analysis N=", nrow(sample)),
                     "变量角色" = paste(capture.output(str(roles)), collapse = "\n"),
                     "清洗规则" = "列名标准化、特殊缺失码转 NA、数值变量 1%/99% 缩尾、去重、删除 Y/X/control 缺失。"))
log_run(args, "02-clean-describe", "ok", c(data_path, dict_path, table1, report), "已真实读取、清洗并生成描述统计。")
